### PR TITLE
[Merged by Bors] - feat(CategoryTheory): natural version of `FullyFaithful.homEquiv`

### PR DESCRIPTION
**Pull request by TwoFX**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Yoneda.lean [hunk 1]
```diff
@@ -751,4 +751,30 @@ lemma yonedaMap_app_apply {Y : C} {X : Cᵒᵖ} (f : X.unop ⟶ Y) :
 
 end
 
+namespace Functor.FullyFaithful
+
+variable {C : Type u₁} [Category.{v₁} C]
+
+/-- Natural version of `FullyFaithful.homEquiv`. -/
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Yoneda.lean:758]:
```suggestion
/-- `FullyFaithful.homEquiv` as a natural isomorphism. -/
```

File: Mathlib/CategoryTheory/Yoneda.lean [hunk 2]
```diff
@@ -751,4 +751,30 @@ lemma yonedaMap_app_apply {Y : C} {X : Cᵒᵖ} (f : X.unop ⟶ Y) :
...
+def homNatIso {D : Type u₂} [Category.{v₂} D] {F : C ⥤ D} (hF : F.FullyFaithful) (X : C) :
+    F.op ⋙ yoneda.obj (F.obj X) ⋙ uliftFunctor.{v₁} ≅ yoneda.obj X ⋙ uliftFunctor.{v₂} :=
+  NatIso.ofComponents
+    (fun Y => Equiv.toIso (Equiv.ulift.trans <| hF.homEquiv.symm.trans Equiv.ulift.symm)) (by
+    intros Y Z f
+    ext ⟨x⟩
+    have : hF.preimage (F.map f.unop ≫ x) = f.unop ≫ hF.preimage x := hF.map_injective (by simp)
+    simp [Equiv.ulift, this])
+
+/-- Natural version of `FullyFaithful.homEquiv`. -/
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Yoneda.lean:768]:
```suggestion
/-- `FullyFaithful.homEquiv` as a natural isomorphism. -/
```

File: Mathlib/CategoryTheory/Yoneda.lean [hunk 3]
```diff
@@ -751,4 +751,30 @@ lemma yonedaMap_app_apply {Y : C} {X : Cᵒᵖ} (f : X.unop ⟶ Y) :
...
+def homNatIso {D : Type u₂} [Category.{v₂} D] {F : C ⥤ D} (hF : F.FullyFaithful) (X : C) :
+    F.op ⋙ yoneda.obj (F.obj X) ⋙ uliftFunctor.{v₁} ≅ yoneda.obj X ⋙ uliftFunctor.{v₂} :=
+  NatIso.ofComponents
+    (fun Y => Equiv.toIso (Equiv.ulift.trans <| hF.homEquiv.symm.trans Equiv.ulift.symm)) (by
+    intros Y Z f
+    ext ⟨x⟩
+    have : hF.preimage (F.map f.unop ≫ x) = f.unop ≫ hF.preimage x := hF.map_injective (by simp)
+    simp [Equiv.ulift, this])
```
[the remark this answers is not in the document]
Comment (TwoFX) [reply Mathlib/CategoryTheory/Yoneda.lean]: Very nice, thanks!

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**